de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/concept/types/structural/level/HierarchyLevelChecker.java

```diff
@@ -0,0 +1,368 @@
+		int minLevelIndex = levels.indexOf(parentLevel);
+
+		// If the level cannot be nested (->repeated) then the element needs to be of
+		// the next level
+		if (!parentLevel.canBeNested()) {
+			minLevelIndex += 1;
+		}
+
+		return minLevelIndex;
+	}
+
+	/**
+	 * Get the minimum level index from the next parent element with level
+	 * 
+	 * @param child
+	 *            the next child with level
+	 * @return the minumum level index
+	 */
+	private int getMaxIndexFromRelevantChild(IBeanStructuralElementInstance child) {
+		IHierarchyLevel childLevel = getLevelOfBean(child);
+
+		int maxLevelIndex = levels.indexOf(childLevel);
+
+		// If the level cannot be nested (->repeated) then the element needs to be of
+		// the previous level from the order
+		if (!childLevel.canBeNested()) {
+			maxLevelIndex -= 1;
+		}
+
+		return maxLevelIndex;
+	}
+
+	/**
+	 * Get the minimum index considering the tree distance of the next relvant child
+	 * with level
+	 * 
+	 * @param child
+	 *            the next relevant child with level - the next relevant child is
+	 *            the one with the least value of the difference from the level
+	 *            index and the tree distance
+	 * @param elementToCheck
+	 *            the element to check
+	 * @return the minimum level index
+	 */
+	private int getMinIndexFromTreeDistanceOfChild(IBeanStructuralElementInstance child,
+			IBeanStructuralElementInstance elementToCheck) {
+		IHierarchyLevel childLevel = getLevelOfBean(child);
+		int treeDistance = getTreeDistance(child, elementToCheck);
+		int minLevel = levels.indexOf(childLevel) - treeDistance;
+
+		// Go through all the levels between the child level and the one with a
+		// distance of the tree depth and check if their optional. If so, then the
+		// minimum applicable level index decreases as we can also skip the level
+		for (int i = levels.indexOf(childLevel) - 1; i >= 0 && i >= levels.indexOf(childLevel) - treeDistance; i--) {
+			if (levels.get(i).isOptional()) {
+				minLevel--;
+			}
+		}
+
+		return minLevel;
+	}
+
+	/**
+	 * Get the maximum level index considering the tree distance to the next parent
+	 * element with level
+	 * 
+	 * @param parent
+	 *            the next parent element with level
+	 * @param elementToCheck
+	 *            the element to check
+	 * @return the maximum level index
+	 */
+	private int getMaxIndexFromTreeDistanceOfParent(IBeanStructuralElementInstance parent,
+			IBeanStructuralElementInstance elementToCheck) {
+		IHierarchyLevel parentLevel = getLevelOfBean(parent);
+		int treeDistance = getTreeDistance(elementToCheck, parent);
+		int maxLevel = levels.indexOf(parentLevel) + treeDistance;
+
+		// Go through all the levels between the parent level and the one with a
+		// distance of the tree depth and check if their optional. If so, then the
+		// maximum applicable level index increases as we can also skip the level
+		for (int i = levels.indexOf(parentLevel) + 1; i <= levels.indexOf(parentLevel) + treeDistance

```

Comment:
Looks like `levels.indexOf(parentLevel) + treeDistance` is `maxLevel`

Reply:
Ah no you modify it in the loop. But I stll think this can be simplified. maybe with an initialMaxLevel.